File: HackerRank/30 days of code/day 22, binary search trees/day_22_binary_search_trees.cpp
#include <iostream>
#include <cstddef>
using namespace std;

class Node
{
public:
    int data;
    Node *left, *right;

    Node(int d) {
        data = d;
        left = right = nullptr;
    }
};

class Solution
{
public:

    Node* insert(Node* root, int data) {

        if (!root)
            return new Node(data);
        else {
            Node* cur;
            if (data <= root->data) {
                cur = insert(root->left, data);
                root->left = cur;
            }
            else {
                cur = insert(root->right, data);
                root->right = cur;
            }
            return root;
        }
    }

    static int get_height(Node *root) {
        static int height {}, maxHeight {};

        if (root->left || root->right) {
            ++height;
            if (root->left)
                get_height(root->left);
            if (root->right)
                get_height(root->right);
            --height;
        }
        else if (height > maxHeight)
            maxHeight = height;
        
        return maxHeight;
    }
};

int main() {
    Solution my_tree;
    Node* root {nullptr};
    int n;
    cin >> n;

    for (int data; cin >> data; root = my_tree.insert(root, data));
    
    cout << Solution::get_height(root);

    return 0;
}

File: HackerRank/30 days of code/day 22, binary search trees/day_22_binary_search_trees.ts
'use strict';

process.stdin.resume();
process.stdin.setEncoding('utf-8');
let inputString: string = '';
let inputLines: string[] = [];
let currentLine: number = 0;
process.stdin.on('data', function(inputStdin: string): void {
    inputString += inputStdin;
});

process.stdin.on('end', function(): void {
    inputLines = inputString.split('\n');
    inputString = '';
    main();
});

function readLine(): string {
    return inputLines[currentLine++];
}

interface Node {
    data: number
    left: Node | null
    right: Node | null
}

class Tree {

    static insert(root : Node, data : number) : Node {
        if (root === null) {
            this.root = new Node(data);
            return this.root;
        }
        
        if (data <= root.data) {
            if (root.left)
                this.insert(root.left, data);
            else
                root.left = new Node(data);
        }
        else {
            if (root.right)
                this.insert(root.right, data);
            else
                root.right = new Node(data);
        }
        
        return this.root;
    }

    static getHeight(root : Node) : number {

        arguments.callee.height = arguments.callee.height || 0
        arguments.callee.maxHeight = arguments.callee.maxHeight || 0

        if (root.left || root.right) {
            arguments.callee.height++
            if (root.left)
                this.getHeight(root.left)
            if (root.right)
                this.getHeight(root.right)
            arguments.callee.height--
        }
        else if (arguments.callee.height > arguments.callee.maxHeight)
            arguments.callee.maxHeight = arguments.callee.height
        
        return arguments.callee.maxHeight
    }
}

function main() {
    
    let root: Node = null;
    
    readLine()
    let values = readLine.split().map(Number);
    
    values.forEach((value) => {root = Tree.insert(root, value)});
    
    console.log(Tree.getHeight(root));
}
